File: sdk/typescript/src/instrument/openai.ts
/**
 * OpenAI 自动埋点。
 *
 * 与 Python SDK 对等：显式包装公开方法，不全局 monkey patch 私有属性。
 * 后者在上游小版本升级时必碎，且会污染同进程内其他库的调用。
 *
 * 用法：
 *   import OpenAI from "openai";
 *   import { wrapOpenAI } from "@ariadne/sdk/instrument/openai";
 *   const client = wrapOpenAI(new OpenAI({ apiKey: "..." }));
 *   await client.chat.completions.create({ model: "gpt-4o", messages: [...] });
 */

import { getClient } from "../client";
import { extractOpenAIUsage, previewMessages, previewOutput } from "./common";

const PROVIDER = "openai";

interface OpenAIRequest {
  model?: string;
  messages?: unknown;
  input?: unknown;
  temperature?: number;
  max_tokens?: number;
  max_completion_tokens?: number;
  stream?: boolean;
}

/**
 * 包装 OpenAI 客户端的 chat.completions.create 方法。
 * 返回原客户端，透明地产生 span。
 */
export function wrapOpenAI<T extends Record<string, unknown>>(client: T): T {
  const completions = (client as unknown as {
    chat?: { completions?: { create?: (...a: unknown[]) => unknown } };
  }).chat?.completions;

  if (completions && typeof completions.create === "function") {
    const originalCreate = completions.create;
    completions.create = function (
      ...args: unknown[]
    ): unknown {
      const request = (args[0] as OpenAIRequest) ?? {};
      return _instrumentCreate(
        () => originalCreate.apply(completions, args),
        request,
        "chat.completions.create",
      );
    };
  }

  const embeddings = (client as unknown as {
    embeddings?: { create?: (...a: unknown[]) => unknown };
  }).embeddings;

  if (embeddings && typeof embeddings.create === "function") {
    const originalEmbed = embeddings.create;
    embeddings.create = function (
      ...args: unknown[]
    ): unknown {
      const request = (args[0] as OpenAIRequest) ?? {};
      return _instrumentCreate(
        () => originalEmbed.apply(embeddings, args),
        request,
        "embeddings.create",
      );
    };
  }

  return client;
}

function _instrumentCreate(
  invoke: () => unknown,
  request: OpenAIRequest,
  operation: string,
): unknown {
  const ariadne = getClient();
  if (!ariadne) return invoke();

  const model = request.model ?? "";
  const span = ariadne.span(`${operation} ${model}`.trim(), {
    kind: "llm",
    operation,
    provider: PROVIDER,
    model,
  });

  span.setAttributes({
    temperature: request.temperature,
    max_tokens: request.max_tokens ?? request.max_completion_tokens,
    stream: request.stream ?? false,
  });

  if (request.messages) {
    span.setInput(previewMessages(request.messages as never));
  } else if (request.input) {
    span.setInput(String(request.input).slice(0, 1024));
  }

  span.enter();

  const result = invoke();

  if (result instanceof Promise) {
    return result.then(
      (resp: unknown) => {
        const r = resp as { model?: string; usage?: unknown };
        if (r?.model) span.setModel(model, { responseModel: r.model });
        const usage = extractOpenAIUsage(r?.usage as never);
        span.setUsage(usage);
        span.setOutput(previewOutput(resp as never));
        span.end();
        return resp;
      },
      (err: Error) => {
        span.recordError(err);
        span.end();
        throw err;
      },
    );
  }

  const r = result as { model?: string; usage?: unknown };
  if (r?.model) span.setModel(model, { responseModel: r.model });
  span.setUsage(extractOpenAIUsage(r?.usage as never));
  span.setOutput(previewOutput(result as never));
  span.end();
  return result;
}

File: src/ariadne_sdk/instrument/openai.py
"""OpenAI 自动埋点。

采用**显式包装公开方法**而非全局 monkey patch 私有属性：
后者在上游小版本升级时必碎，且会污染同进程内其他库的调用。

用法：
    from openai import OpenAI
    from ariadne_sdk.instrument.openai import wrap_openai

    client = wrap_openai(OpenAI())
    client.chat.completions.create(...)   # 自动产生 span
"""

from __future__ import annotations

import functools
from typing import Any, TypeVar

from ariadne_sdk.client import get_client
from ariadne_sdk.instrument.common import (
    extract_openai_usage,
    preview_messages,
    preview_output,
)

T = TypeVar("T")

_PROVIDER = "openai"


def _instrument_create(original: Any, operation: str) -> Any:
    @functools.wraps(original)
    def wrapper(*args: Any, **kwargs: Any) -> Any:
        client = get_client()
        if client is None:
            return original(*args, **kwargs)

        model = str(kwargs.get("model", ""))
        span = client.span(
            f"{operation} {model}".strip(),
            kind="llm",
            operation=operation,
            provider=_PROVIDER,
            model=model,
        )
        with span:
            span.set_attributes(
                {
                    "temperature": kwargs.get("temperature"),
                    "max_tokens": kwargs.get("max_tokens")
                    or kwargs.get("max_completion_tokens"),
                    "stream": kwargs.get("stream", False),
                }
            )
            if messages := kwargs.get("messages"):
                span.set_input(preview_messages(messages))
            elif input_text := kwargs.get("input"):
                span.set_input(preview_output(input_text))

            response = original(*args, **kwargs)

            # 流式响应的 usage 要等迭代结束才有，M1 不拆流；标注后跳过用量
            if kwargs.get("stream"):
                span.set_attributes({"usage_unavailable": "stream"})
                return response

            usage = extract_openai_usage(response)
            span.set_usage(**usage)
            actual_model = getattr(response, "model", "") or model
            span.set_model(model, response_model=str(actual_model), provider=_PROVIDER)
            span.set_output(preview_output(response))
            if finish := _finish_reason(response):
                span.set_attributes({"finish_reason": finish})
            return response

    return wrapper


def _finish_reason(response: Any) -> str:
    choices = getattr(response, "choices", None) or []
    if choices:
        return str(getattr(choices[0], "finish_reason", "") or "")
    return ""


def wrap_openai(client: T) -> T:
    """就地包装 OpenAI 客户端的 create 方法，返回同一实例。

    幂等：重复包装同一客户端不会产生嵌套 span。
    """
    targets = (
        (("chat", "completions"), "chat"),
        (("embeddings",), "embeddings"),
        (("responses",), "chat"),
    )

    for path, operation in targets:
        node: Any = client
        for attr in path:
            node = getattr(node, attr, None)
            if node is None:
                break
        if node is None:
            continue

        create = getattr(node, "create", None)
        if create is None or getattr(create, "_ariadne_wrapped", False):
            continue

        wrapped = _instrument_create(create, operation)
        wrapped._ariadne_wrapped = True
        try:
            node.create = wrapped
        except (AttributeError, TypeError):
            # 某些版本用 __slots__ 或只读属性，跳过而非崩掉业务
            continue

    return client
